`typeparamref` in a member's doc comment which refers to the containing type's type parameter raises an error if in an included XML file

**Version Used**: 2.3.1

**Steps to Reproduce**:

Consider the following C# file:
```csharp
class Foo<T>
{
    /// <include file='Foo.xml' path='docs/*'/>
    public void DoWork()
    { }
}
```

`Foo.xml` is:
```xml
<?xml version="1.0" encoding="utf-8" ?>
<docs>
  <summary>
    <typeparamref name="T" />
  </summary>
</docs>
```

When compiled this produces the following error:
```
Foo.cs(3,9,3,48): warning CS1735: XML comment on 'Foo<T>.DoWork()' has a typeparamref tag for 'T', but there is no type parameter by that name
```

However, when the doc comment is moved inline, there is no error:
```csharp
class Foo<T>
{
    /// <summary>
    /// <typeparamref name="T" />
    /// </summary>
    public void DoWork()
    { }
}
```

--- a/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
+++ b/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
@@ -600,7 +600,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                                 }
                                 break;
                         }
-                    } while (isTypeParameterRef && (currentSymbol = currentSymbol.ContainingSymbol) != null);
+                        currentSymbol = currentSymbol.ContainingSymbol;
+                    } while (isTypeParameterRef && !(currentSymbol is null));
                 }
 
                 return binder;

--- a/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
+++ b/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
@@ -220,12 +220,15 @@ namespace Microsoft.CodeAnalysis.CSharp
                             if (ElementNameIs(element, DocumentationCommentXmlNames.ParameterElementName) ||
                                 ElementNameIs(element, DocumentationCommentXmlNames.ParameterReferenceElementName))
                             {
-                                BindName(attribute, originatingSyntax, isParameter: true);
+                                BindName(attribute, originatingSyntax, isParameter: true, isTypeRef: false);
                             }
-                            else if (ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterElementName) ||
-                                ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterReferenceElementName))
+                            else if (ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterElementName))
                             {
-                                BindName(attribute, originatingSyntax, isParameter: false);
+                                BindName(attribute, originatingSyntax, isParameter: false, isTypeRef: false);
+                            }
+                            else if(ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterReferenceElementName))
+                            {
+                                BindName(attribute, originatingSyntax, isParameter: false, isTypeRef: true);
                             }
                         }
                     }
@@ -514,7 +517,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 crefDiagnostics.Free();
             }
 
-            private void BindName(XAttribute attribute, CSharpSyntaxNode originatingSyntax, bool isParameter)
+            private void BindName(XAttribute attribute, CSharpSyntaxNode originatingSyntax, bool isParameter, bool isTypeRef)
             {
                 XmlNameAttributeSyntax attrSyntax = ParseNameAttribute(attribute.ToString(), attribute.Parent.Name.LocalName);
 
@@ -529,7 +532,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     "Why are we processing a documentation comment that is not attached to a member declaration?");
 
                 DiagnosticBag nameDiagnostics = DiagnosticBag.GetInstance();
-                Binder binder = MakeNameBinder(isParameter, _memberSymbol, _compilation);
+                Binder binder = MakeNameBinder(isParameter, isTypeRef, _memberSymbol, _compilation);
                 DocumentationCommentCompiler.BindName(attrSyntax, binder, _memberSymbol, ref _documentedParameters, ref _documentedTypeParameters, nameDiagnostics);
                 RecordBindingDiagnostics(nameDiagnostics, sourceLocation); // Respects DocumentationMode.
                 nameDiagnostics.Free();
@@ -538,7 +541,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             // NOTE: We're not sharing code with the BinderFactory visitor, because we already have the
             // member symbol in hand, which makes things much easier.
-            private static Binder MakeNameBinder(bool isParameter, Symbol memberSymbol, CSharpCompilation compilation)
+            private static Binder MakeNameBinder(bool isParameter, bool isTypeRef, Symbol memberSymbol, CSharpCompilation compilation)
             {
                 Binder binder = new BuckStopsHereBinder(compilation);
 
@@ -576,23 +579,27 @@ namespace Microsoft.CodeAnalysis.CSharp
                 }
                 else
                 {
-                    switch (memberSymbol.Kind)
+                    bool continueSearch = true;
+                    for (var currentSymbol = memberSymbol; currentSymbol != null && continueSearch; currentSymbol = currentSymbol.ContainingSymbol, continueSearch = isTypeRef)
                     {
-                        case SymbolKind.NamedType: // Includes delegates.
-                        case SymbolKind.ErrorType:
-                            NamedTypeSymbol typeSymbol = (NamedTypeSymbol)memberSymbol;
-                            if (typeSymbol.Arity > 0)
-                            {
-                                binder = new WithClassTypeParametersBinder(typeSymbol, binder);
-                            }
-                            break;
-                        case SymbolKind.Method:
-                            MethodSymbol methodSymbol = (MethodSymbol)memberSymbol;
-                            if (methodSymbol.Arity > 0)
-                            {
-                                binder = new WithMethodTypeParametersBinder(methodSymbol, binder);
-                            }
-                            break;
+                        switch (currentSymbol.Kind)
+                        {
+                            case SymbolKind.NamedType: // Includes delegates.
+                            case SymbolKind.ErrorType:
+                                NamedTypeSymbol typeSymbol = (NamedTypeSymbol)currentSymbol;
+                                if (typeSymbol.Arity > 0)
+                                {
+                                    binder = new WithClassTypeParametersBinder(typeSymbol, binder);
+                                }
+                                break;
+                            case SymbolKind.Method:
+                                MethodSymbol methodSymbol = (MethodSymbol)currentSymbol;
+                                if (methodSymbol.Arity > 0)
+                                {
+                                    binder = new WithMethodTypeParametersBinder(methodSymbol, binder);
+                                }
+                                break;
+                        }
                     }
                 }
 

--- a/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
+++ b/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
@@ -220,15 +220,15 @@ namespace Microsoft.CodeAnalysis.CSharp
                             if (ElementNameIs(element, DocumentationCommentXmlNames.ParameterElementName) ||
                                 ElementNameIs(element, DocumentationCommentXmlNames.ParameterReferenceElementName))
                             {
-                                BindName(attribute, originatingSyntax, isParameter: true, isTypeRef: false);
+                                BindName(attribute, originatingSyntax, isParameter: true, isTypeParameterRef: false);
                             }
                             else if (ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterElementName))
                             {
-                                BindName(attribute, originatingSyntax, isParameter: false, isTypeRef: false);
+                                BindName(attribute, originatingSyntax, isParameter: false, isTypeParameterRef: false);
                             }
                             else if(ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterReferenceElementName))
                             {
-                                BindName(attribute, originatingSyntax, isParameter: false, isTypeRef: true);
+                                BindName(attribute, originatingSyntax, isParameter: false, isTypeParameterRef: true);
                             }
                         }
                     }
@@ -517,7 +517,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 crefDiagnostics.Free();
             }
 
-            private void BindName(XAttribute attribute, CSharpSyntaxNode originatingSyntax, bool isParameter, bool isTypeRef)
+            private void BindName(XAttribute attribute, CSharpSyntaxNode originatingSyntax, bool isParameter, bool isTypeParameterRef)
             {
                 XmlNameAttributeSyntax attrSyntax = ParseNameAttribute(attribute.ToString(), attribute.Parent.Name.LocalName);
 
@@ -532,7 +532,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     "Why are we processing a documentation comment that is not attached to a member declaration?");
 
                 DiagnosticBag nameDiagnostics = DiagnosticBag.GetInstance();
-                Binder binder = MakeNameBinder(isParameter, isTypeRef, _memberSymbol, _compilation);
+                Binder binder = MakeNameBinder(isParameter, isTypeParameterRef, _memberSymbol, _compilation);
                 DocumentationCommentCompiler.BindName(attrSyntax, binder, _memberSymbol, ref _documentedParameters, ref _documentedTypeParameters, nameDiagnostics);
                 RecordBindingDiagnostics(nameDiagnostics, sourceLocation); // Respects DocumentationMode.
                 nameDiagnostics.Free();
@@ -541,7 +541,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             // NOTE: We're not sharing code with the BinderFactory visitor, because we already have the
             // member symbol in hand, which makes things much easier.
-            private static Binder MakeNameBinder(bool isParameter, bool isTypeRef, Symbol memberSymbol, CSharpCompilation compilation)
+            private static Binder MakeNameBinder(bool isParameter, bool isTypeParameterRef, Symbol memberSymbol, CSharpCompilation compilation)
             {
                 Binder binder = new BuckStopsHereBinder(compilation);
 
@@ -579,8 +579,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 }
                 else
                 {
-                    bool continueSearch = true;
-                    for (var currentSymbol = memberSymbol; currentSymbol != null && continueSearch; currentSymbol = currentSymbol.ContainingSymbol, continueSearch = isTypeRef)
+                    var currentSymbol = memberSymbol;
+                    do
                     {
                         switch (currentSymbol.Kind)
                         {
@@ -600,7 +600,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                                 }
                                 break;
                         }
-                    }
+                    } while (isTypeParameterRef && (currentSymbol = currentSymbol.ContainingSymbol) != null);
                 }
 
                 return binder;

--- a/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
+++ b/src/Compilers/CSharp/Portable/Compiler/DocumentationCommentCompiler.IncludeElementExpander.cs
@@ -226,7 +226,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                             {
                                 BindName(attribute, originatingSyntax, isParameter: false, isTypeParameterRef: false);
                             }
-                            else if(ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterReferenceElementName))
+                            else if (ElementNameIs(element, DocumentationCommentXmlNames.TypeParameterReferenceElementName))
                             {
                                 BindName(attribute, originatingSyntax, isParameter: false, isTypeParameterRef: true);
                             }
@@ -579,7 +579,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 }
                 else
                 {
-                    var currentSymbol = memberSymbol;
+                    Symbol currentSymbol = memberSymbol;
                     do
                     {
                         switch (currentSymbol.Kind)

--- a/src/Compilers/CSharp/Test/Symbol/DocumentationComments/DocumentationCommentCompilerTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/DocumentationComments/DocumentationCommentCompilerTests.cs
@@ -3782,6 +3782,167 @@ partial class C
             Assert.Equal(string.Format(expectedTemplate, TestHelpers.AsXmlCommentText(xmlFilePath)), actual);
         }
 
+        [Fact]
+        [WorkItem(21348, "https://github.com/dotnet/roslyn/issues/21348")]
+        public void IncludedName_TypeParamAndTypeParamRefHandling()
+        {
+            var xml = @"
+<root>
+    <target>
+    Included section
+    <summary>
+      See <typeparam/>.
+      See <typeparam name=""X""/>.
+      See <typeparam name=""Y""/>.
+      See <typeparam name=""XY""/>.
+    </summary>
+    <remarks></remarks>
+    </target>
+    <target>
+    Included section
+    <summary>
+      See <typeparamref/>.
+      See <typeparamref name=""X""/>.
+      See <typeparamref name=""Y""/>.
+      See <typeparamref name=""XY""/>.
+    </summary>
+    <remarks></remarks>
+    </target>
+</root>
+";
+
+            var xmlFile = Temp.CreateFile(extension: ".xml").WriteAllText(xml);
+            var xmlFilePath = xmlFile.Path;
+
+            var includeElementTemplate = @"<include file='{0}' path='//target'/>";
+            string includeElement = string.Format(includeElementTemplate, xmlFilePath);
+
+            var sourceTemplate = @"
+
+/// {0}
+class OuterClass<X>
+{{
+    /// {0}
+    class InnerClass<Y>
+    {{
+        /// {0}
+        public void Foo() {{}}
+    }}
+}}
+";
+            var comp = CreateCompilationUtil(string.Format(sourceTemplate, includeElement));
+
+            var actual = GetDocumentationCommentText(comp, expectedDiagnostics: new[] {
+                // (3,5): warning CS1711: XML comment has a typeparam tag for 'Y', but there is no type parameter by that name
+                // /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("Y").WithLocation(3, 5),
+                // (3,5): warning CS1711: XML comment has a typeparam tag for 'XY', but there is no type parameter by that name
+                // /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("XY").WithLocation(3, 5),
+                // (3,5): warning CS1735: XML comment on 'OuterClass<X>' has a typeparamref tag for 'Y', but there is no type parameter by that name
+                // /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamRefTag, includeElement).WithArguments("Y", "OuterClass<X>").WithLocation(3, 5),
+                // (3,5): warning CS1735: XML comment on 'OuterClass<X>' has a typeparamref tag for 'XY', but there is no type parameter by that name
+                // /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamRefTag, includeElement).WithArguments("XY", "OuterClass<X>").WithLocation(3, 5),
+                // (6,9): warning CS1711: XML comment has a typeparam tag for 'X', but there is no type parameter by that name
+                //     /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("X").WithLocation(6, 9),
+                // (6,9): warning CS1711: XML comment has a typeparam tag for 'XY', but there is no type parameter by that name
+                //     /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("XY").WithLocation(6, 9),
+                // (6,9): warning CS1735: XML comment on 'OuterClass<X>.InnerClass<Y>' has a typeparamref tag for 'XY', but there is no type parameter by that name
+                //     /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamRefTag, includeElement).WithArguments("XY", "OuterClass<X>.InnerClass<Y>").WithLocation(6, 9),
+                // (9,13): warning CS1711: XML comment has a typeparam tag for 'X', but there is no type parameter by that name
+                //         /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("X").WithLocation(9, 13),
+                // (9,13): warning CS1711: XML comment has a typeparam tag for 'Y', but there is no type parameter by that name
+                //         /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("Y").WithLocation(9, 13),
+                // (9,13): warning CS1711: XML comment has a typeparam tag for 'XY', but there is no type parameter by that name
+                //         /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamTag, includeElement).WithArguments("XY").WithLocation(9, 13),
+                // (9,13): warning CS1735: XML comment on 'OuterClass<X>.InnerClass<Y>.Foo()' has a typeparamref tag for 'XY', but there is no type parameter by that name
+                //         /// <include file='b16c2dc7f738.xml' path='//target'/>
+                Diagnostic(ErrorCode.WRN_UnmatchedTypeParamRefTag, includeElement).WithArguments("XY", "OuterClass<X>.InnerClass<Y>.Foo()").WithLocation(9, 13)
+            });
+            var expected = (@"
+<?xml version=""1.0""?>
+<doc>
+    <assembly>
+        <name>Test</name>
+    </assembly>
+    <members>
+        <member name=""T:OuterClass`1"">
+            <target>
+    Included section
+    <summary>
+      See <typeparam />.
+      See <typeparam name=""X"" />.
+      See <typeparam name=""Y"" />.
+      See <typeparam name=""XY"" />.
+    </summary>
+    <remarks />
+    </target><target>
+    Included section
+    <summary>
+      See <typeparamref />.
+      See <typeparamref name=""X"" />.
+      See <typeparamref name=""Y"" />.
+      See <typeparamref name=""XY"" />.
+    </summary>
+    <remarks />
+    </target>
+        </member>
+        <member name=""T:OuterClass`1.InnerClass`1"">
+            <target>
+    Included section
+    <summary>
+      See <typeparam />.
+      See <typeparam name=""X"" />.
+      See <typeparam name=""Y"" />.
+      See <typeparam name=""XY"" />.
+    </summary>
+    <remarks />
+    </target><target>
+    Included section
+    <summary>
+      See <typeparamref />.
+      See <typeparamref name=""X"" />.
+      See <typeparamref name=""Y"" />.
+      See <typeparamref name=""XY"" />.
+    </summary>
+    <remarks />
+    </target>
+        </member>
+        <member name=""M:OuterClass`1.InnerClass`1.Foo"">
+            <target>
+    Included section
+    <summary>
+      See <typeparam />.
+      See <typeparam name=""X"" />.
+      See <typeparam name=""Y"" />.
+      See <typeparam name=""XY"" />.
+    </summary>
+    <remarks />
+    </target><target>
+    Included section
+    <summary>
+      See <typeparamref />.
+      See <typeparamref name=""X"" />.
+      See <typeparamref name=""Y"" />.
+      See <typeparamref name=""XY"" />.
+    </summary>
+    <remarks />
+    </target>
+        </member>
+    </members>
+</doc>
+        ").Trim();
+            Assert.Equal(expected, actual);
+        }
+
         #endregion Included names
 
         #endregion Include

